libraries/Microsoft.Bot.Builder.Core.State/FileSystemStateStorageProvider.cs

```diff
@@ -0,0 +1,330 @@
+                using (var jsonReader = new JsonTextReader(textReader))
+                {
+                    var data = (JObject)(await JObject.ReadFromAsync(jsonReader));
+
+                    return new FileSystemStateStorageEntry(stateNamespace, key, storageEntryFileInfo.FullName, data);
+                }
+            }
+            catch (DirectoryNotFoundException)
+            {
+                // TODO: log
+                return default(IStateStorageEntry);
+            }
+            catch (FileNotFoundException)
+            {
+                // TODO: log
+                return default(IStateStorageEntry);
+            }
+        }
+
+        private sealed class FileSystemStateStorageEntry : DeferredValueStateStorageEntry
+        {
+            private JObject _data;
+
+            public FileSystemStateStorageEntry(string stateNamespace, string key, string filePath) : base(stateNamespace, key)
+            {
+                FilePath = filePath ?? throw new ArgumentNullException(nameof(filePath));
+            }
+
+            public FileSystemStateStorageEntry(string stateNamespace, string key, string filePath, JObject data) : this(stateNamespace, key, filePath)
+            {
+                _data = data ?? throw new ArgumentNullException(nameof(data));
+
+                ETag = data.Value<JObject>("@@botFramework").Value<string>("eTag");
+            }
+
+            public string FilePath { get; }
+
+            public async Task WriteToFile()
+            {
+                tryAgain:
+                try
+                {
+                    using (var fileStream = new FileStream(FilePath, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.None, bufferSize: 8192, options: FileOptions.Asynchronous | FileOptions.RandomAccess))
+                    {
+                        if (fileStream.Length > 0)
+                        {
+                            await EnusureETagsMatch(fileStream);
+                        }
+
+                        fileStream.Position = 0;
+
+                        _data = RawValue != null ? JObject.FromObject(RawValue) : new JObject();
+
+                        _data["@@botFramework"] = new JObject()
+                        {
+                            ["rawNamespace"] = Namespace,
+                            ["rawKey"] = Key,
+                            ["eTag"] = Guid.NewGuid().ToString("N")
+                        };
+
+                        using (var textWriter = new StreamWriter(fileStream, Encoding.UTF8))
+                        using (var jsonWriter = new JsonTextWriter(textWriter))
+                        {
+                            await _data.WriteToAsync(jsonWriter);
+                        }
+                    }
+                }
+                catch (DirectoryNotFoundException)
+                {
+                    Directory.CreateDirectory(Path.GetDirectoryName(FilePath));
+
+                    goto tryAgain;
```

Comment:
Or, ya know, a proper `RetryPolicy`? 😜

I've added a counter in for now, but this is another one of those things that needs to be a bullet point on the hardening issue we put in for this.

Reply:
I do want to point out that it would take some extraordinary circumstances to end up in an infinite loop because the exception handler creates the directory and if that were to fail it would throw on that call and never hit the `goto`. So the only thing that would cause an infinite loop would be some weird scenario where creating the directory succeeds, but then that very next ms somehow the directory kept getting deleted again.

Anyway, it should absolutely have a proper retry policy passed into this class for this kind of stuff, but I've got the retry counter in there for now.